File: voice_assistant/audio/vad.py
"""Reusable energy-based voice activity detection (VAD).

Stateful segmenter: feed it fixed-size int16 frames one at a time; it returns
a completed utterance (np.int16, shape (N, channels)) when speech ends, else
None. The noise floor adapts continuously, so it works as an always-on stream
processor (GUI) as well as a one-shot recorder (CLI).
"""

from collections import deque

import numpy as np

from voice_assistant.config import config


def rms(frame: np.ndarray) -> float:
    x = frame.astype(np.float64)
    return float(np.sqrt(np.mean(x * x))) if x.size else 0.0


class VADSegmenter:
    def __init__(
        self,
        sample_rate: int,
        frame_ms: int,
        silence_tail: float,
        max_seconds: float,
        threshold_factor: float,
        threshold_floor: float,
        preroll_s: float = 0.15,
    ):
        self.sample_rate = sample_rate
        self.frame_len = int(sample_rate * frame_ms / 1000)
        dt = frame_ms / 1000
        self.tail_frames = max(1, int(silence_tail / dt))
        self.max_frames = int(max_seconds / dt)
        self.factor = threshold_factor
        self.floor = threshold_floor
        self.preroll: deque = deque(maxlen=max(1, int(preroll_s / dt)))
        self.reset(full=True)

    @classmethod
    def from_config(cls) -> "VADSegmenter":
        return cls(
            config.sample_rate,
            config.vad_frame_ms,
            config.vad_silence_tail,
            config.vad_max_seconds,
            config.vad_threshold_factor,
            config.vad_threshold_floor,
        )

    def reset(self, full: bool = False) -> None:
        self.collecting = False
        self.buf: list[np.ndarray] = []
        self.voiced_run = 0
        self.silence_run = 0
        self.preroll.clear()
        if full:
            self.noise_floor = self.floor / max(self.factor, 1e-6)

    @property
    def is_speaking(self) -> bool:
        return self.collecting

    @property
    def threshold(self) -> float:
        return max(self.noise_floor * self.factor, self.floor)

    def feed(self, frame: np.ndarray) -> np.ndarray | None:
        level = rms(frame)
        if not self.collecting:
            # Adapt the noise floor only while no speech is in progress.
            self.noise_floor = 0.95 * self.noise_floor + 0.05 * level

        threshold = self.threshold
        if not self.collecting:
            self.preroll.append(frame)
            if level > threshold:
                self.voiced_run += 1
                if self.voiced_run >= 2:  # need 2 voiced frames to avoid clicks
                    self.collecting = True
                    self.buf = list(self.preroll)
                    self.silence_run = 0
            else:
                self.voiced_run = 0
            return None

        self.buf.append(frame)
        if level <= threshold:
            self.silence_run += 1
            if self.silence_run >= self.tail_frames:
                return self._finish()
        else:
            self.silence_run = 0
        if len(self.buf) >= self.max_frames:
            return self._finish()
        return None

    def _finish(self) -> np.ndarray:
        audio = np.concatenate(self.buf, axis=0)
        self.reset()  # keep the adapted noise_floor
        return audio

File: voice_assistant/web/src/lib/vad.js
// Browser energy-based VAD — mirrors voice_assistant/audio/vad.py (VADSegmenter).
// Feed it a normalized RMS level (0..1) per analysis frame; it returns
// 'start' when speech begins, 'end' when it ends, or null otherwise.

export class BrowserVAD {
  constructor({
    floor = 0.015,
    factor = 3,
    silenceTail = 0.8,
    maxSeconds = 15,
    frameSec = 1 / 60,
    voicedToStart = 2,
  } = {}) {
    this.floor = floor
    this.factor = factor
    this.tailFrames = Math.max(1, Math.round(silenceTail / frameSec))
    this.maxFrames = Math.round(maxSeconds / frameSec)
    this.voicedToStart = voicedToStart
    this.reset(true)
  }

  reset(full = false) {
    this.collecting = false
    this.voicedRun = 0
    this.silenceRun = 0
    this.frames = 0
    if (full) this.noiseFloor = this.floor / Math.max(this.factor, 1e-6)
  }

  get threshold() {
    return Math.max(this.noiseFloor * this.factor, this.floor)
  }

  /** level: normalized RMS 0..1. boost: multiply threshold (e.g. during playback). */
  feed(level, boost = 1) {
    if (!this.collecting) {
      this.noiseFloor = 0.95 * this.noiseFloor + 0.05 * level
    }
    const thr = this.threshold * boost

    if (!this.collecting) {
      if (level > thr) {
        this.voicedRun += 1
        if (this.voicedRun >= this.voicedToStart) {
          this.collecting = true
          this.frames = 0
          this.silenceRun = 0
          return 'start'
        }
      } else {
        this.voicedRun = 0
      }
      return null
    }

    this.frames += 1
    if (level <= thr) {
      this.silenceRun += 1
      if (this.silenceRun >= this.tailFrames) {
        this.reset()
        return 'end'
      }
    } else {
      this.silenceRun = 0
    }
    if (this.frames >= this.maxFrames) {
      this.reset()
      return 'end'
    }
    return null
  }
}
